js/src/logger.ts

```diff
@@ -1642,6 +1655,87 @@ function validateAndSanitizeExperimentLogFullArgs(
   return event;
 }
 
+class ObjectFetcher<RecordType> {
+  private excludeFields: Set<string> = new Set(["_object_delete"]);
+  private _fetchedData: any[] | undefined = undefined;
+
+  constructor(
+    private objectType: "dataset" | "experiment",
+    private jsonFields: Set<string>,
+    private pinnedVersion: string | undefined
+  ) {}
+
+  public get id(): Promise<string> {
+    throw new Error("ObjectFetcher subclasses must have an 'id' attribute");
+  }
+
+  protected async getState(): Promise<BraintrustState> {
+    throw new Error("ObjectFetcher subclasses must have a 'getState' method");
+  }
+
+  async *fetch(): AsyncGenerator<RecordType> {
+    const records = await this.fetchedData();
+    for (const record of records) {
+      yield Object.fromEntries(
+        Object.entries(record)
+          .filter(([k, _]) => !this.excludeFields.has(k))
+          .map(([k, v]) => [
+            k,
+            this.jsonFields.has(k) && typeof v === "string" ? JSON.parse(v) : v,
+          ])
+      ) as RecordType;
+    }
+    this.clearCache();
+  }
+
+  [Symbol.iterator]() {
+    return this.fetch();
+  }
+
+  async fetchedData() {
+    if (this._fetchedData === undefined) {
+      const state = await this.getState();
+      const resp = await state.logConn().get(`object/${this.objectType}`, {
+        id: await this.id,
+        fmt: "json",
+        version: this.pinnedVersion,
+      });
+
+      const text = await resp.text();
+      this._fetchedData = text
+        .split("\n")
+        .filter((x: string) => x.trim() !== "")
+        .map((x: string) => JSON.parse(x));
+    }
+    return this._fetchedData || [];
+  }
+
+  clearCache() {
+    this._fetchedData = undefined;
+  }
+
+  public async version() {
```

Comment:
super nit: we're being a bit inconsistent about marking methods public here.

Reply:
we can go through this later (i just copy/pasted as they currently are)